luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,350 @@
+			shieldDamage = 0,                                              -- This stores the value of damages populated in ShieldPreDamaged(), then applied in GameFrame() all at once
+			shieldWeaponNumber = -1,                                       -- This is replaced with the real shieldWeaponNumber as soon as the shield is damaged
+			downtime = shieldUnitDefs[unitDefID].customParams.shield_downtime or 8, -- Defined in unitdef.customparams with a default fallback value
+			downtimeReset = 0,
+			shieldCoverageChecked = false,									-- Used to prevent expensive unit coverage checks being performed more than once per cycle
+			radius = shieldUnitDefs[unitDefID].customParams.shield_radius
+		}
+	end
+
+	-- Increases performance by reducing global unitDefID lookups
+	unitDefIDCache[unitID] = unitDefID
+end
+
+function gadget:UnitDestroyed(unitID)
+	shieldUnitsData[unitID] = nil
+	unitDefIDCache[unitID] = nil
+end
+
+function gadget:ProjectileCreated(proID, proOwnerID, weaponDefID)
+	-- Increases performance by reducing global projectileDefID lookups
+	projectileDefIDCache[proID] = weaponDefID
+end
+
+function gadget:ProjectileDestroyed(proID)
+	projectileDefIDCache[proID] = nil
+	projectileShieldHitCache[proID] = nil
+end
+
+local function triggerDowntime(unitID, weaponNum)
+	local shieldData = shieldUnitsData[unitID]
+
+	-- This method is used for mobile units with shields such as evocom cortex commander. This is far less efficient, but for smaller unit counts is OK.
+	spSetUnitShieldRechargeDelay(unitID, weaponNum, 120)
+
+	spSetUnitShieldState(unitID, weaponNum, false)
+	shieldData.downtimeReset = gameSeconds + shieldData.downtime
+	shieldData.shieldEnabled = false
+end
+
+local function removeCoveredUnits(shieldUnitID)
+	for unitID, shieldList in pairs(shieldedUnits) do
+		if shieldList[shieldUnitID] then
+			shieldList[shieldUnitID] = nil
+		end
+	end
+end
+
+local function setCoveredUnits(shieldUnitID)
+	local shieldData = shieldUnitsData[shieldUnitID]
+
+	if not shieldData then
+		return
+	else
+		removeCoveredUnits(shieldUnitID)
+		local x, y, z = spGetUnitPosition(shieldUnitID, true)
+		local unitsTable = spGetUnitsInSphere(x, y, z, (shieldData.radius - 10))
+
+		for _, unitID in ipairs(unitsTable) do
+			if shieldedUnits[unitID] then
+				shieldedUnits[unitID][shieldUnitID] = true
+			else
+				shieldedUnits[unitID] = {}
+				shieldedUnits[unitID][shieldUnitID] = true
+			end
+		end
+
+		shieldData.shieldCoverageChecked = true
+	end
+end
+
+local function shieldNegatesDamageCheck(unitID, unitTeam, attackerID, attackerTeam)
+	if shieldedUnits[unitID] and allTeams[unitTeam] ~= allTeams[attackerTeam] and next(shieldedUnits[unitID]) then
+		if not shieldedUnits[attackerID] or (not next(shieldedUnits[attackerID]) and next(shieldedUnits[unitID])) then
+			return true
+		end
+
+		for subKey in pairs(shieldedUnits[unitID]) do
```

Comment:
```suggestion
		for shieldUnitID, _ in pairs(shieldedUnits[unitID]) do
```
`subKey` is not descriptive of what this variable represents, all iterator variables should be explicitly in code, even if unused